src/goto-programs/rw_set.cpp

```diff
@@ -36,7 +36,9 @@ void rw_sett::read_write_rec(
     const symbolt *symbol = ns.lookup(symbol_expr.get_identifier());
     if(symbol)
     {
-      if(!symbol->static_lifetime && !expr.type().is_pointer())
+      if(
+        !symbol->static_lifetime &&
+        expr.type().subtype().id() != "dereferenced_symbol")
```

Comment:
Why should we ingore `` expr.type().subtype().id() != "dereferenced_symbol"''? can you please add some explanation here?

Reply:
We're not ignoring `dereferenced_symbol` here, we're actually adding a data races check assertion to the symbol it's `dereferenced_symbol`.

For example:
```
void* t1(void *arg) {
    int* shared = (int*) arg;
    *shared = 4;            
}
int main() {
    pthread_t id1;
    int shared_data = 0;

    pthread_create(&id1, NULL, t1, &shared_data);

    return 0;
}
```
The `shared_data` pointed to here is a local variable, and we might **need a way** to get it past the filter in this code.
So I added a subtype named `dereferenced_symbol` to all dereferenced `expr`.

Reply:
I see.

What would be the difference between `dereference` (already in ESBMC) and `dereferenced_symbol`?

Reply:
Well, here dereference means `*shared`, and dereferenced_symbol means `shared_data`.

Reply:
Thanks for this explanation. Can you improve the documentation in the code?

Reply:
Sure! I will improve them, thanks for your review.

Reply:
We shouldn't be introducing new types, as it will probably crash esbmc if it ever reaches the backend. To add a new type, we should add it to both irep1 and irep2 + implement the appropriate migration.

So what you want to do is to get the content of a global variable? You can probably do that by checking if the static/global flags are set on the symbol

Reply:
> We shouldn't be introducing new types, as it will probably crash esbmc if it ever reaches the backend. To add a new type, we should add it to both irep1 and irep2 + implement the appropriate migration.

I thought about this, but none of the exprs are passed back to the backend, which only uses this to decide to add assertions to check for data races.

> So what you want to do is to get the content of a global variable? You can probably do that by checking if the static/global flags are set on the symbol

The tricky part is that these symbols are not global variables, as shown above, `shared_data` is only a local variable, but we need to deal with that.

Reply:
Since we need it as the object of the assertion, it would be violated if it appeared in a different thread. Dereferencing may have different variable names, but the object they point to remains the same.

Reply:
I didn't understand, can you give an example?

Reply:
```
#include <pthread.h>

void* t1(void *arg) {
    int* shared1 = (int*) arg;
    *shared1 = 4;                 //race here

    return NULL;
}

void* t2(void *arg) {
    int* shared2 = (int*) arg;
    *shared2 = 2;                //race here

    return NULL;
}

int main() {
    pthread_t id1, id2;
    int shared_data = 0;

    pthread_create(&id1, NULL, t1, &shared_data);
    pthread_create(&id2, NULL, t2, &shared_data);

    pthread_join(id1, NULL);
    pthread_join(id2, NULL);

    return 0;
}
```
Sure, as shown in the example above, we are creating two threads to run the functions t1 and t2. As you can see in the two functions we have variables `shared1` and `shared2`, they have different variable names but they both point to `shared_data`. 

Our current approach to data race detection is to use assert to determine if the same identifier exists in multiple threads (e.g., `assert! shared1`), we would not get the correct result. So we need to dereference it and we get the symbol - `shared_data`, So whether it's `shared1` or `shared2`, we use `shared_data` to add assert (`assert! shared_data`)

Reply:
and why we create an `assert !shared1` on the `*shared1 = 4;`?

It should clearly be `assert !*shared1`. Is esbmc ignoring the dereference?

Reply:
In the current master goto functions are as follows:
```
        *shared1=4;
        // 247 file main6.c line 5 column 5 function t1
        ASSERT !tmp_c:main6.c@48@F@t1@shared1 // R/W data race on c:main6.c@48@F@t1@shared1
```

```
        *shared2=2;
        // 257 file main6.c line 12 column 5 function t2
        ASSERT !tmp_c:main6.c@168@F@t2@shared2 // R/W data race on c:main6.c@168@F@t2@shared2
```
esbmc does not ignore dereference, but it only takes the identifer of the symbol to add assertions.
https://github.com/esbmc/esbmc/blob/27c6bbbbefd636956d0dd4cd28238e1fa4cfd109/src/goto-programs/rw_set.cpp#L60-L66